src/Raven.Server/Utils/AffinityHelper.cs

```diff
@@ -0,0 +1,201 @@
+﻿using System;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using Raven.Client.Util;
+using Sparrow.Binary;
+using Sparrow.Collections;
+using Sparrow.Logging;
+using Sparrow.Platform;
+using Sparrow.Server.Platform.Posix;
+
+namespace Raven.Server.Utils
+{
+    public class AffinityHelper
+    {
+        private static readonly EasyReaderWriterLock _affinityLocker = new EasyReaderWriterLock();
+        private static readonly ConcurrentSet<PoolOfThreads.PooledThread> _customAffinityThreads = new ConcurrentSet<PoolOfThreads.PooledThread>();
+        private static readonly Logger _logger = LoggingSource.Instance.GetLogger<AffinityHelper>("Server");
+
+        public static void SetProcessAffinity(Process process, int cores, long? processAffinityMask, out long currentlyAssignedCores)
+        {
+            currentlyAssignedCores = Bits.NumberOfSetBits(process.ProcessorAffinity.ToInt64());
+            if (currentlyAssignedCores == cores)
+            {
+                // we already set the correct number of assigned cores
+                return;
+            }
+
+            var bitMask = 1L;
+            if (processAffinityMask == null)
+            {
+                for (var i = 0; i < cores; i++)
+                {
+                    bitMask |= 1L << i;
+                }
+            }
+            else if (Bits.NumberOfSetBits(processAffinityMask.Value) > cores)
+            {
+                var affinityMask = processAffinityMask.Value;
+                var bitNumber = 0;
+                while (cores > 0)
+                {
+                    if ((affinityMask & 1) != 0)
+                    {
+                        bitMask |= 1L << bitNumber;
+                        cores--;
+                    }
+
+                    affinityMask = affinityMask >> 1;
+                    bitNumber++;
+                }
+            }
+            else
+            {
+                bitMask = processAffinityMask.Value;
+            }
+
+            _affinityLocker.EnterWriteLock();
+
+            using (_affinityLocker.EnterWriteLock())
+            {
+                process.ProcessorAffinity = new IntPtr(bitMask);
+
+                // changing the process affinity resets the thread affinity
+                // we need to change the custom affinity threads as well
+                foreach (var pooledThread in _customAffinityThreads)
```

Comment:
I'm pretty sure that once we set it at the process level, it applies to the threads too

Reply:
It does. But it's changing it to the Process affinity and we need it to be 1 less (by default).

Reply:
Okay